### PR TITLE
RPPL-2578: unit test's added for thunder_broker.rs 

### PR DESCRIPTION
**Pull request by nnaveen979**
## What

unit test's added for thunder_broker.rs 

## Why

verify the functionality and improve code coverage.

## How

added new test cases.

## Test

Generated report locally to check the coverage improvement.

## Checklist

- [+] I have self-reviewed this PR
- [ ] I have added tests that prove the feature works or the fix is effective

### CODE REVIEW

File: core/main/src/broker/thunder_broker.rs [hunk 1]
```diff
@@ -330,3 +330,197 @@ impl EndpointBroker for ThunderBroker {
...
+
+    #[tokio::test]
+    async fn test_thunderbroker_subscribe() {
+        let (tx, mut _rx) = mpsc::channel(1);
+        let (sender, _rec) = mpsc::channel(1);
+        let send_data = vec![WSMockData::get(json!({"key":"value"}).to_string())];
+
+        let thndr_broker = get_thunderbroker(tx, send_data, sender.clone(), false).await;
+
+        let request = BrokerRequest {
```
Comment (Vinodsathyaseelan) [inline core/main/src/broker/thunder_broker.rs:480]: This test appears to be not correct. I do not seen any subscription method as part of the rpc request.  Please add additional use case like.
1) Subscribe to even1
2)Generate event1
3)Assert that the event1 has been received by the subscriber.
4)Unsubscribe event1
5) Invoke step 2 and confirm that no more event notification to the subscriber.

**Review by Vinodsathyaseelan — CHANGES_REQUESTED** [review]
.

File: core/main/src/broker/thunder_broker.rs [hunk 2]
```diff
@@ -330,3 +330,291 @@ impl EndpointBroker for ThunderBroker {
...
+
+    #[tokio::test]
+    async fn test_thunderbroker_subscribe_unsubscribe() {
+        let (tx, mut _rx) = mpsc::channel(1);
+        let (sender, mut rec) = mpsc::channel(1);
+        let send_data = vec![WSMockData::get(
+            json!({
+                "jsonrpc": "2.0",
+                "method": "AcknowledgeChallenge.onRequestChallenge",
+                "params": {
```
[the remark this answers is not in the document]
Comment (nnaveen979) [reply core/main/src/broker/thunder_broker.rs:480]: i have added listen param as true for subscribe, please review these.

File: core/main/src/broker/thunder_broker.rs [hunk 3]
```diff
@@ -330,3 +330,285 @@ impl EndpointBroker for ThunderBroker {
...
+        if let Some(broker_output) = v {
+            let data = broker_output
+                .data
+                .result
+                .expect("No result in response data");
+            let key_value = data.get("key").expect("Key not found in response data");
+            let key_str = key_value.as_str().expect("Value is not a string");
+            assert_eq!(key_str, "value");
+        } else {
+            panic!("Received None instead of a valid response");
```
Comment (brendanobra) [inline core/main/src/broker/thunder_broker.rs:468]: change this to an assert, i.e. `assert!(broker_output.is_some())`

File: core/main/src/broker/thunder_broker.rs [hunk 4]
```diff
@@ -330,3 +330,285 @@ impl EndpointBroker for ThunderBroker {
...
+        thndr_broker.sender.send(request).await.unwrap();
+
+        let _v = tokio::time::timeout(Duration::from_secs(2), rec.recv())
+            .await
+            .expect("Timeout while waiting for response");
+
+        assert!(_rx.recv().await.unwrap());
+
+        //invoke prepare_request()
+        let rpc_request = BrokerRequest {
```
Comment (brendanobra) [inline core/main/src/broker/thunder_broker.rs:400]: might make sense to factor `BrokerRequest` creation into a `fn` that takes an parameters that vary.

File: core/main/src/broker/thunder_broker.rs [hunk 5]
```diff
@@ -330,3 +330,285 @@ impl EndpointBroker for ThunderBroker {
...
+        };
+
+        match thndr_broker.prepare_request(&rpc_request) {
+            Ok(requests) => {
+                for request in requests {
+                    println!("Prepared request: {}", request);
+                }
+            }
+            Err(e) => {
+                error!("Failed to prepare request: {:?}", e);
```
Comment (brendanobra) [inline core/main/src/broker/thunder_broker.rs:418]: Should this fail the test? should it be an assert?

File: core/main/src/broker/thunder_broker.rs [hunk 6]
```diff
@@ -330,3 +330,285 @@ impl EndpointBroker for ThunderBroker {
...
+    #[tokio::test]
+    async fn test_thunderbroker_get_cleaner() {
+        let (tx, mut _rx) = mpsc::channel(1);
+        let (sender, _rec) = mpsc::channel(1);
+        let send_data = vec![WSMockData::get(json!({"key":"value"}).to_string())];
+
+        let thndr_broker = get_thunderbroker(tx, send_data, sender, false).await;
+
+        let cleaner = thndr_broker.get_cleaner();
+        assert!(cleaner.cleaner.is_some());
```
Comment (brendanobra) [inline core/main/src/broker/thunder_broker.rs:432]: Can this be shortened to : `assert!(thndr_broker.get_cleaner().is_some())` ?

File: core/main/src/broker/thunder_broker.rs [hunk 7]
```diff
@@ -330,3 +330,285 @@ impl EndpointBroker for ThunderBroker {
...
+            "jsonrpc": "2.0",
+            "result": {
+                "key": "value"
+            }
+        });
+
+        let callback = BrokerCallback {
+            sender: sender.clone(),
+        };
+        ThunderBroker::handle_jsonrpc_response(response.to_string().as_bytes(), callback);
```
Comment (brendanobra) [inline core/main/src/broker/thunder_broker.rs:453]: create callback inline on line 453 to cut down on code?

Comment (nnaveen979) [reply core/main/src/broker/thunder_broker.rs]: review comments fixed please have a look

File: core/main/src/broker/thunder_broker.rs (hunk 5 above)
Comment (nnaveen979) [reply core/main/src/broker/thunder_broker.rs]: review comments fixed please have a look

File: core/main/src/broker/thunder_broker.rs (hunk 4 above)
Comment (nnaveen979) [reply core/main/src/broker/thunder_broker.rs]: review comments fixed please have a look

File: core/main/src/broker/thunder_broker.rs (hunk 3 above)
Comment (nnaveen979) [reply core/main/src/broker/thunder_broker.rs]: review comments fixed please have a look

File: core/main/src/broker/thunder_broker.rs (hunk 1 above)
Comment (nnaveen979) [reply core/main/src/broker/thunder_broker.rs]: review comments fixed please review them.

File: core/main/src/broker/thunder_broker.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (nnaveen979) [reply core/main/src/broker/thunder_broker.rs:506]: fixed review comments please review these.

**Review by Vinodsathyaseelan — APPROVED** [review]

**Review by tfarz123 — APPROVED** [review]

**Pull request merged**